src/aggregators/coevolve.jl

```diff
@@ -0,0 +1,206 @@
+                                                   Tuple{typeof(u), typeof(p), typeof(t)}}
+    if (variable_jumps !== nothing) && !isempty(variable_jumps)
+        affects! = [AffectWrapper((integrator) -> (c.affect!(integrator); nothing))
+                    for c in variable_jumps]
+        rates = [RateWrapper(c.rate) for c in variable_jumps]
+        lrates = Any[RateWrapper(c.lrate) for c in variable_jumps]
+        urates = Any[RateWrapper(c.urate) for c in variable_jumps]
+        Ls = Any[RateWrapper(c.L) for c in variable_jumps]
+    else
+        affects! = Vector{AffectWrapper}()
+        rates = Vector{RateWrapper}()
+        lrates = Vector{RateWrapper}()
+        urates = Vector{RateWrapper}()
+        Ls = Vector{RateWrapper}()
+    end
+    cur_rates = nothing
+    sum_rate = nothing
+    next_jump = 0
+    next_jump_time = typemax(t)
+    CoevolveJumpAggregation(next_jump, next_jump_time, end_time, cur_rates, sum_rate,
+                            ma_jumps, rates, affects!, save_positions, rng;
+                            u = u,
+                            dep_graph = dep_graph,
+                            lrates = lrates, urates = urates, Ls = Ls)
+end
+
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    p.end_time = integrator.sol.prob.tspan[2]
+    fill_rates_and_get_times!(p, u, params, t)
+    generate_jumps!(p, integrator, u, params, t)
+    nothing
+end
+
+# execute one jump, changing the system state
+function execute_jumps!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    # execute jump
+    u = update_state!(p, integrator, u, params, t)
+    # update current jump rates and times
+    update_dependent_rates!(p, u, params, t)
+    nothing
+end
+
+# calculate the next jump / jump time
+function generate_jumps!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    p.next_jump_time, p.next_jump = top_with_handle(p.pq)
+    nothing
+end
+
+@inline function update_state!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    @unpack ma_jumps, next_jump = p
+    num_majumps = get_num_majumps(ma_jumps)
+    if next_jump <= num_majumps
+        if u isa SVector
+            integrator.u = executerx(u, next_jump, ma_jumps)
+        else
+            @inbounds executerx!(u, next_jump, ma_jumps)
+        end
+    else
+        idx = next_jump - num_majumps
+        @inbounds p.affects![idx](integrator)
+    end
+    p.prev_jump = next_jump
+    return integrator.u
+end
+
+######################## SSA specific helper routines ########################
+function update_dependent_rates!(p::CoevolveJumpAggregation, u, params, t)
+    @inbounds deps = p.dep_gr[p.next_jump]
+    @unpack end_time, pq = p
+    for (ix, i) in enumerate(deps)
+        ti = next_time(p, u, params, t, i, end_time)
+        update!(pq, i, ti)
+    end
+    nothing
+end
+
+function get_rates(p::CoevolveJumpAggregation, i, u)
```

Comment:
Are you saying that we should move the logic in this function to inside of `next_time` instead of calling it from there?

In `NRM` the `cur_rates` vector is updated inside of the function `update_dependent_rates!` by calling the function `calculate_jump_rate`, reproduced below: 
```julia
@inline function calculate_jump_rate(ma_jumps, num_majumps, rates, u, params, t, rx)
    if rx <= num_majumps
        return evalrxrate(u, rx, ma_jumps)
    else
        @inbounds return rates[rx - num_majumps](u, params, t)
    end
end
```

This function also branches to compute the rates for either `MassActionJump` or `ConstantRateJump`. So, in a sense it does very similar to `get_rates` in this PR.

One of the main differences is that `get_rates` is not inlined. I'm not sure if this could cause a performance penalty.

Reply:
Resolving this conversation as it is reflected in other sections of this PR.